### PR TITLE
[refactor](Nereids): avoid useless groupByColStats Map

### PR DESCRIPTION
**Pull request by jackwener**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jackwener** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 53.07 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17170271716 Bytes

**Comment by jackwener** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.82 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          30.1 seconds inserted 10000000 Rows, about 332K ops/s
 storage size: 17167895337 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.44 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162454863 Bytes

**Comment by jackwener** [comment]
run buildp0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 1]
```diff
@@ -687,37 +687,32 @@ private Statistics computeLimit(Limit limit) {
...
+                    return colStats.isUnKnown() ? -1 : colStats.ndv;
+                })
+                .sorted(Comparator.reverseOrder())
+                .collect(Collectors.toList());
+        if (groupByExpressions.isEmpty()) {
+            return 1;
         }
-        int groupByCount = groupByExpressions.size();
-        if (groupByColStats.values().stream().anyMatch(ColumnStatistic::isUnKnown)) {
+        if (groupByNdvs.stream().anyMatch(ndv -> ndv == -1)) {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:703]: isUnknown is not equal to "ndv==-1".

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 2]
```diff
@@ -687,37 +687,32 @@ private Statistics computeLimit(Limit limit) {
...
-                colStats = ExpressionEstimation.estimate(groupByExpr, childStats);
-            }
-            groupByColStats.put(groupByExpr, colStats);
+        List<Double> groupByNdvs = groupByExpressions.stream()
+                .map(groupByExpr -> {
+                    ColumnStatistic colStats = childStats.findColumnStatistics(groupByExpr);
+                    if (colStats == null) {
+                        colStats = ExpressionEstimation.estimate(groupByExpr, childStats);
+                    }
+                    return colStats.isUnKnown() ? -1 : colStats.ndv;
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:696]: never set ndv to -1

**Comment by englefly** [comment]
what is the benifit of this refactor?

**Comment by jackwener** [comment]
> what is the benifit of this refactor?

avoid to generate useless HashMap. It will cost some time.

**Comment by jackwener** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.23 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17167067837 Bytes

**Review by starocean999 — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**